### DISCUSSION TITLE
Know device IP address after connection stablished

### DISCUSSION THREAD

**Discussion by andycolone** [Hieromon/AutoConnect#495] [How can I?]
Hello there,

Great library and great job! 
First attempt to use it here. I was able to use the Getting Started guide to compile and upload the basic sketch.

I can also use my Android phone to be able to access my home ap. Unfortunately, I cannot know the IP address assigned by my router as the captive portal disappears after joining my home wifi network. Which seems to be different to what states in the [docs](https://hieromon.github.io/AutoConnect/gettingstarted.html#connection-establishment).

I know the IP address because it is printed out on the serial monitor.

How do I know the assigned IP address after joining my home wifi?

Thanks much!

**Comment by Hieromon** [comment]
@andycolone, This is somewhat strange. 
> I cannot know the IP address assigned by my router as the captive portal disappears after joining my home wifi network. Which seems to be different to what states in the [docs](https://hieromon.github.io/AutoConnect/gettingstarted.html#connection-establishment).

The IP address returned on the following screen is the IP address of your ESP module when it had established a connection with the access point. It is assigned by a WiFi router with DHCP capability. Once you reset the ESP module here, the automatic connection after reboot will have that IP address enabled. So you should be able to access it from your browser with that IP address.

Does the phenomenon you encountered that there is no response when you access its IP address from your browser?

<img width="240px" src="https://hieromon.github.io/AutoConnect/images/established.png">

Alternatively, if you cannot know the IP address later, you can take the following measures.

1. Use mDNS responder
    With the mDNS responder, you can access the ESP module by name instead of IP address. How to incorporate mDNS into your sketch is described in the document [here](https://hieromon.github.io/AutoConnect/adnetwork.html#combination-with-mdns).

2. Use a Network IP Scanner
    Android apps have tools that scan the IP addresses of local network segments, and there are many similar apps. They usually have a WiFi LAN scanner with device name resolution. You will know the IP address of the ESP module using the network IP scanner tools with your Android that is connected to the same network as the ESP module.

**Comment by andycolone** [comment]
Thanks a lot for your quick response. 

Yes, my home wifi router has DHCP capabilities. 

As I had the ESP32 board connected through USB, I was able to know the IP and connect to the board using my computer (including the captive portal on ip_address/_ac endpoint).

The issue was related to the second (local IP address) screen never showing up as shown in the docs. 

Not connected to this, but I also tried setting the SSID/PK using the config with no luck either. 

May it be something related to libraries for ESP32?

Here is my sketch

```
#include <WiFi.h>
#include <WebServer.h>

//#include <ESPmDNS.h>
#include <AutoConnect.h>

WebServer Server;

AutoConnect Portal(Server);
AutoConnectConfig Config;


void rootPage() {
  char content[] = "Hello, world";
  Server.send(200, "text/plain", content);
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println("starting-up");

  //Config.retainPortal = true;   // Keep the captive portal open.
  //Config.immediateStart = true;
  //Config.preserveAPMode = true;
  //Config.autoReset = false;     // Not reset the module even by intentional disconnection using AutoConnect menu.
  Config.apid = "ap_portal";
  Config.psk  = "123";
  Portal.config( Config );
  

  Server.on("/", rootPage);
  
  if ( Portal.begin() ) {
    Serial.println("WiFi connected: " + WiFi.localIP().toString());
     //if ( MDNS.begin("dnsName") ) {
     // MDNS.addService("http", "tcp", 80);
    //}
  }

}

void loop() {
    Portal.handleClient();
}
```

**Reply by Hieromon** [reply]
@andycolone 
> The issue was related to the second (local IP address) screen never showing up as shown in the docs.

Does your mention mean that even if you successfully connect to the AP, your Android will not transition to the screen described above?
If so, what do you see on your Android screen? (like forever waiting for a response, or an unreachable message)

**EDIT**
The following guidance may help you solve this problem.

The ESP32 has a WiFi implementation with two different roles. One realizes a Soft AP and the other constitutes a WiFi station. The TCP stacks of Soft AP and the WiFi station are independent and have separate IP addresses.

In the situation where you open a captive portal with the ESP module, the IP address of the ESP module is in the following state.

![OpenCP](https://user-images.githubusercontent.com/12591771/172306050-bdb17525-7e94-42f9-b602-2c94ae18aba7.png)

The connection request is issued from the web page displayed on your android. (i.e. HTTP GET) The response to this GET request is pending for the duration of the connection attempt.
If the connection is successful, the status of the peripherals around your android will look as follows:

![Conn](https://user-images.githubusercontent.com/12591771/172306465-b268d268-d8f4-4299-adb4-2b3f6644400b.png)

If Android does not receive the "**c. Responds a connected page**" response above, this is not normal behavior. The HTTP session for the "**a. HTTP - GET**" as a connection request has not been completed. It is "Connection Reset by peer" or "Timeout". This can occur due to signal loss in the WIFI_AP or if your android left the ESP32 WIFI_AP and connects an existing home access point automatically.

![LoseConn](https://user-images.githubusercontent.com/12591771/172310226-8dcb1ade-cfe3-412a-bed2-5f3f825f33de.svg)

**Comment by andycolone** [comment]
Hello Hieromon,

> Does your mention mean that even if you successfully connect to the AP, your Android will not transition to the screen described above?
> If so, what do you see on your Android screen? (like forever waiting for a response, or an unreachable message)

After the connection is established, it transitions to a "page not found", but it is quite quicliky. I am almost not able to see it.
After that, I am connected to the WIFI_AP but no captive portal showing the current IP address on the local network.

Is there any way to make sure the Android is redirected to a page showing the current IP address?

I am assuming everyone using the library will need that address in order to do OTA or any other functionality besides connecting the LAN.
 
Really appreciate your help!

**Reply by Hieromon** [reply]
@andycolone 
> After the connection is established, it transitions to a "page not found"

Your android may have already broken the connection with SoftAP in the ESP module.
Can you enable debug logging in AutoConnect to get logs output to the serial monitor? See [here](https://hieromon.github.io/AutoConnect/adothers.html#debug-print) in the documentation to enable debug logging.
Also, build the sketch with the Arduino ESP32 core debug level set to 5. Refer to the description [here](https://docs.espressif.com/projects/arduino-esp32/en/latest/guides/tools_menu.html?highlight=core%20debug%20level#core-debug-level) for how to set the debug level in the Arduino IDE.

If you find the **AP_STADISCONNECTED** or the **Connection reset by peer** in the log, your Android has left the ESP32 SoftAP immediately after the WIFI_STA connection. This is not an AutoConnect problem, but a problem with the Android device settings or the ESP module.
Also, check your android IP address at this time. If that IP address is the LAN segment of your HOME AP, your android is leaving the ESP32 SoftAP and had connected automatically to the HOME AP. (i.e., your android itself is switching the access point) I have explained this in the diagram above.

> After that, I am connected to the WIFI_AP but no captive portal showing the current IP address on the local network.

It is normal behavior, AutoConnect will close the captive portal once it has established a WIFI_STA connection with the default settings.

> I am assuming everyone using the library will need that address in order to do OTA or any other functionality besides connecting the LAN.

Even Android devices can transit to a page that displays the IP address of the ESP module after establishing a connection in the captive portal. The problem that this topic shows is that the AutoConnect page transition does not work correctly on your Android. My Android device does not reproduce this phenomenon.
To proceed with the problem diagnosis, you need to get the debug log mentioned above.

**EDIT: More information**

I have since explored other helpful information and it seems to be discussing an issue related to SoftAP connections on Android 10. This phenomenon is not reproduced on my Android device but may occur on another device's Android implementation.
https://github.com/espressif/arduino-esp32/issues/4423

In the middle of that discussion, I had the following comments:

> By disabling the cellular data on your Android 10 phone and connect it to the WiFi access point of the ESP32, it should work fine.

https://github.com/espressif/arduino-esp32/issues/4423#issuecomment-712179095
https://github.com/espressif/arduino-esp32/issues/4423#issuecomment-712181179

Can you try this on android as I would like to know if the situation improves with this solution?

**EDIT: Workaround worth trying**

It may be worth trying the following changes, although the information is uncertain. Follow the example below to change the IP address of the SoftAP that AutoConnect will launch. The default SoftAP IP assigned by AutoConnect, **`172.217.28.1`** is the public range and is compatible with Android 7 implementations. This setting will cause the captive portal to pop up automatically when Android connects to the ESP32 SoftAP. The following test code dares to disguise its IP as a Google DNS IP. 

I still can't say for sure, but I believe this issue is related to the security enhancements in Android 10 and later.  And this test code may lose the captive portal auto popup. In that case, please go directly to **`http://8.8.4.4/_ac`** with the browser of your Android device. This temporary workaround is less convenient, but I believe it will help us locate the problem.

Thank you for your contribution.

```cpp
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

WebServer Server;
AutoConnect Portal(Server);
AutoConnectConfig Config;

void rootPage() {
  char content[] = "Hello, world";
  Server.send(200, "text/plain", content);
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  Serial.println("starting-up");

  Config.apid = "ap_portal";
  Config.psk  = "12345678"; // Password must be least 8 characters.

  // Disguise the SoftAPIP as a Google DNS
  Config.apip = IPAddress(8, 8, 4, 4);
  Config.gateway = IPAddress(8, 8, 4, 4);

  Portal.config(Config);  

  Server.on("/", rootPage);
  
  if (Portal.begin()) {
    Serial.println("WiFi connected: " + WiFi.localIP().toString());
  }
}

void loop() {
  Portal.handleClient();
}
```
